### PR TITLE
Address mapping in EditEngine

### PR DESCRIPTION
**Pull request by catsuryuu**

### CODE REVIEW

File: include/util/edit.h [hunk 1]
```diff
@@ -21,6 +21,8 @@
 #include <QMap>
 #include <QVector>
 
+#include <memory>
```
Comment (mkow) [inline include/util/edit.h]: Please move it to the group with stdlib

File: include/util/edit.h [hunk 2]
```diff
@@ -52,12 +56,27 @@ class EditEngine {
   data::BinData originalBytesValues(size_t offset, size_t size) const;
 
  private:
+  struct EditNode {
+    std::shared_ptr<data::BinData> fragment_;
```
Comment (mkow) [inline include/util/edit.h]: Why shared_ptr? It doesn't seem to be shared with anything outside of this object.

File: include/util/edit.h [hunk 3]
```diff
@@ -52,12 +56,27 @@ class EditEngine {
...
+    size_t offset_;
+
+    explicit EditNode(const data::BinData& fragment, size_t offset)
+        : fragment_(std::make_shared<data::BinData>(fragment)),
+          offset_(offset) {}
+    explicit EditNode(size_t offset) : fragment_(nullptr), offset_(offset) {}
```
Comment (mkow) [inline include/util/edit.h]: Simpler constructor first.

File: src/util/edit.cc [hunk 4]
```diff
@@ -109,6 +109,23 @@ size_t EditEngine::undo() {
 }
 
 void EditEngine::applyChanges() {
+  for (auto it = address_mapping_.cbegin(); it != address_mapping_.cend();
```
Comment (mkow) [inline src/util/edit.cc]: Add a comment what this loop does.

File: src/util/edit.cc [hunk 5]
```diff
@@ -33,6 +36,46 @@ void EditEngine::changeBytes(size_t pos, const data::BinData& bytes,
...
+    } else {
+      it->fragment_->setData(it->offset_ + (pos - it.key()), bytes.size(),
+                             bytes);
+    }
+  } else {
+    it = next_it;
+    ++next_it;
+    while (next_it != address_mapping_.cend() && next_it.key() < end_pos) {
+      address_mapping_.erase(it);
+      it = next_it;
```
Comment (mkow) [inline src/util/edit.cc]:
```C++
address_mapping_.erase(it);
it = next_it;
```
-->
```C++
it = address_mapping_.erase(it);
```

File: src/util/edit.cc [hunk 6]
```diff
@@ -139,9 +183,19 @@ void EditEngine::clear() {
...
+uint64_t EditEngine::byteValue(size_t pos) const {
+  auto next_it = address_mapping_.upperBound(pos);
+  assert(next_it != address_mapping_.cbegin());
+  auto it = next_it;
+  --it;
+
+  if (it->fragment_ == nullptr) {
+    return original_data_->binData().element64(it->offset_ + (pos - it.key()));
+  } else {
+    return it->fragment_->element64(it->offset_ + (pos - it.key()));
```
Comment (mkow) [inline src/util/edit.cc]: extract duplicated `it->offset_ + (pos - it.key())` to a variable with a descriptive name

File: src/util/edit.cc [hunk 7]
```diff
@@ -150,28 +204,71 @@ uint64_t EditEngine::byteValue(size_t byte_pos) const {
...
+    result.setData(0, size,
+                   getDataFromEditNode(it.value(), pos - it.key(), size));
+  } else {
+    size_t size_to_write = next_it.key() - pos;
+    result.setData(
+        0, size_to_write,
+        getDataFromEditNode(it.value(), pos - it.key(), size_to_write));
+    size_t bytes_written = size_to_write;
+
+    it = next_it;
```
Comment (mkow) [inline src/util/edit.cc]: Use `++it` instead

File: src/util/edit.cc [hunk 8]
```diff
@@ -33,6 +36,46 @@ void EditEngine::changeBytes(size_t pos, const data::BinData& bytes,
...
+            end_pos, EditNode(nullptr, it->offset_ + (end_pos - it.key())));
+      }
+      address_mapping_.insert(pos, EditNode(bytes));
+      // this node can overwrite `*it` and that's OK
+    } else {
+      it->fragment_->setData(it->offset_ + (pos - it.key()), bytes.size(),
+                             bytes);
+    }
+  } else {
+    it = next_it;
```
Comment (mkow) [inline src/util/edit.cc]: `it = next_it;` --> `++it;`

File: src/util/edit.cc [hunk 9]
```diff
@@ -150,28 +204,71 @@ uint64_t EditEngine::byteValue(size_t byte_pos) const {
...
+    size_t bytes_written = size_to_write;
+
+    it = next_it;
+    ++next_it;
+    while (next_it != address_mapping_.cend() && next_it.key() < end_pos) {
+      size_to_write = next_it.key() - it.key();
+      result.setData(bytes_written, size_to_write,
+                     getDataFromEditNode(it.value(), 0, size_to_write));
+      bytes_written += size_to_write;
+      it = next_it;
```
Comment (mkow) [inline src/util/edit.cc]: `++it;`

File: include/util/edit.h [hunk 10]
```diff
@@ -79,6 +79,8 @@ class EditEngine {
 
   data::BinData getDataFromEditNode(const EditNode& edit_node, size_t offset,
                                     size_t size) const;
+  void trySquashWithPrev(QMap<size_t, EditNode>::iterator it);
```
Comment (mkow) [inline include/util/edit.h]: pass arg by const ref

File: include/util/edit.h [hunk 11]
```diff
@@ -79,6 +79,8 @@ class EditEngine {
...
+  void trySquash(QMap<size_t, EditNode>::iterator it);
```
Comment (mkow) [inline include/util/edit.h]: ditto

File: src/util/edit.cc [hunk 12]
```diff
@@ -192,6 +196,41 @@ data::BinData EditEngine::getDataFromEditNode(const EditNode& edit_node,
   }
 }
 
+/*
+ * Warning: it can destruct iterator `it`
```
Comment (mkow) [inline src/util/edit.cc]: This is not destruction but invalidation.

**Comment by mkow** [comment]
Please add more comments to the code, especially at places where iterators are used.

File: src/util/edit.cc (hunk 6 above)
Comment (catsuryuu) [reply src/util/edit.cc]: Now in lines 136-139.

File: include/util/edit.h (hunk 10 above)
Comment (catsuryuu) [reply include/util/edit.h]: DONE

File: include/util/edit.h (hunk 11 above)
Comment (catsuryuu) [reply include/util/edit.h]: DONE

File: src/util/edit.cc (hunk 12 above)
Comment (catsuryuu) [reply src/util/edit.cc]: DONE

File: include/util/edit.h (hunk 3 above)
Comment (catsuryuu) [reply include/util/edit.h]: OUTDATED

File: include/util/edit.h (hunk 1 above)
Comment (catsuryuu) [reply include/util/edit.h]: DONE

File: src/util/edit.cc (hunk 5 above)
Comment (catsuryuu) [reply src/util/edit.cc]: DONE

File: src/util/edit.cc (hunk 8 above)
Comment (catsuryuu) [reply src/util/edit.cc]: DONE

File: src/util/edit.cc (hunk 7 above)
Comment (catsuryuu) [reply src/util/edit.cc]: DONE

File: src/util/edit.cc (hunk 9 above)
Comment (catsuryuu) [reply src/util/edit.cc]: DONE

File: include/util/edit.h [hunk 13]
```diff
@@ -60,22 +60,26 @@ class EditEngine {
     std::shared_ptr<data::BinData> fragment_;
     size_t offset_;
 
-    explicit EditNode(const data::BinData& fragment, size_t offset)
-        : fragment_(std::make_shared<data::BinData>(fragment)),
-          offset_(offset) {}
-    explicit EditNode(size_t offset) : fragment_(nullptr), offset_(offset) {}
+    explicit EditNode(std::shared_ptr<data::BinData> fragment, size_t offset)
```
[the remark this answers is not in the document]
Comment (catsuryuu) [reply include/util/edit.h]: OUTDATED, but DONE

File: src/util/edit.cc (hunk 6 above)
Comment (catsuryuu) [reply src/util/edit.cc]: DONE

File: src/util/edit.cc (hunk 4 above)
Comment (catsuryuu) [reply src/util/edit.cc]: I'll comment the code after implementing insert and delete.

**Pull request merged**